lib/node.js

```diff
@@ -81,36 +75,181 @@ Node.prototype._loadBitcoind = function(config) {
+      }
+
+      done(null, ancestorHash);
+    });
+  });
+};
+
+/**
+ * This function will attempt to rewind the chain to the common ancestor
+ * between the current chain and a forked block.
+ * @param {Block} block - The new tip that forks the current chain.
+ * @param {Function} done - A callback function that is called when complete.
+ */
+Node.prototype._syncBitcoindRewind = function(block, done) {
+
+  var self = this;
+
+  self._syncBitcoindAncestor(block, function(err, ancestorHash) {
+
+    // Rewind the chain to the common ancestor
+    async.whilst(
+      function() {
+        // Wait until the tip equals the ancestor hash
+        return self.chain.tip.hash !== ancestorHash;
+      },
+      function(removeDone) {
+
+        var tip = self.chain.tip;
+
+        self.getBlock(tip.prevHash, function(err, previousTip) {
+          if (err) {
+            removeDone(err);
+          }
+
+          // Undo the related indexes for this block
+          self.db._onChainRemoveBlock(tip, function(err) {
+            if (err) {
+              return removeDone(err);
+            }
+
+            // Set the new tip
+            delete self.chain.tip.__transactions;
+            previousTip.__height = self.chain.tip.__height - 1;
+            self.chain.tip = previousTip;
+            self.chain.saveMetadata();
+            self.chain.emit('removeblock', tip);
+            removeDone();
+          });
+
+        });
+
+      }, done
+    );
+  });
+};
+
+/**
+ * This function will synchronize additional indexes for the chain based on
+ * the current active chain in the bitcoin daemon. In the event that there is
+ * a reorganization in the daemon, the chain will rewind to the last common
+ * ancestor and then resume syncing.
+ */
 Node.prototype._syncBitcoind = function() {
   var self = this;
 
+  if (self.bitcoindSyncing) {
+    return;
+  }
+
+  if (!self.chain.tip) {
+    return;
+  }
+
+  self.bitcoindSyncing = true;
+
   log.info('Starting Bitcoind Sync');
 
-  var info = self.bitcoind.getInfo();
   var height;
 
   async.whilst(function() {
-    if (self.syncStrategy !== Node.SYNC_STRATEGIES.BITCOIND) {
-      log.info('Stopping Bitcoind Sync');
-      return false;
-    }
     height = self.chain.tip.__height;
-    return height < info.blocks;
-  }, function(next) {
+    return height < self.bitcoindHeight;
+  }, function(done) {
     self.bitcoind.getBlock(height + 1, function(err, blockBuffer) {
       if (err) {
-        return next(err);
+        return done(err);
+      }
+      var block = self.Block.fromBuffer(blockBuffer);
+      if (block.prevHash === self.chain.tip.hash) {
+
+        // This block appends to the current chain tip and we can
+        // immediately add it to the chain and create indexes.
+
+        // Populate height
+        block.__height = self.chain.tip.__height + 1;
+
+        // Create indexes
+        self.db._onChainAddBlock(block, function(err) {
```

Comment:
Added the prevHash cache again, good catch. I wonder, is this cache needed with the addition of `bitcoind.getBlockIndex()`?

Calling `db._onChainAddBlock()` directly helps since in this case most of the functionality that was in-between is handled by the bitcoin daemon or not being utilized:
- merkle root validation
- attempting to write the block (since the db method `putBlock()` was replaced with an empty function, and the chain method `_writeBlock()` was nearly empty)
- updating the weight, the db method `db._updateWeight` was replaced with an empty function. Populating `__weight` (chain work) isn't needed since consensus in handled by bitcoind.
- validating proof-of-work as all consensus validation is handled by bitcoind, and getting blocks via height
- the block queue wasn't utilized when syncing from bitcoind